File: qt/gen_qcolorspace.cpp
// +build ignore

#include <QByteArray>
#include <QColorSpace>
#include <QColorTransform>
#include <QList>
#include <QPointF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qcolorspace.h>
#include "gen_qcolorspace.h"

#ifndef _Bool
#define _Bool bool
#endif

void _GUID_Delete(_GUID* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<_GUID*>( self );
	} else {
		delete self;
	}
}

void type_info_Delete(type_info* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<type_info*>( self );
	} else {
		delete self;
	}
}

QColorSpace* QColorSpace_new() {
	return new QColorSpace();
}

QColorSpace* QColorSpace_new2(NamedColorSpace namedColorSpace) {
	return new QColorSpace(namedColorSpace);
}

QColorSpace* QColorSpace_new3(QPointF* whitePoint, TransferFunction transferFunction) {
	return new QColorSpace(*whitePoint, transferFunction);
}

QColorSpace* QColorSpace_new4(QPointF* whitePoint, struct miqt_array /* of uint16_t */  transferFunctionTable) {
	QList<uint16_t> transferFunctionTable_QList;
	transferFunctionTable_QList.reserve(transferFunctionTable.len);
	uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
	for(size_t i = 0; i < transferFunctionTable.len; ++i) {
		transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
	}
	return new QColorSpace(*whitePoint, transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new5(Primaries primaries, TransferFunction transferFunction) {
	return new QColorSpace(primaries, transferFunction);
}

QColorSpace* QColorSpace_new6(Primaries primaries, float gamma) {
	return new QColorSpace(primaries, static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new7(Primaries primaries, struct miqt_array /* of uint16_t */  transferFunctionTable) {
	QList<uint16_t> transferFunctionTable_QList;
	transferFunctionTable_QList.reserve(transferFunctionTable.len);
	uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
	for(size_t i = 0; i < transferFunctionTable.len; ++i) {
		transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
	}
	return new QColorSpace(primaries, transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new8(QPointF* whitePoint, QPointF* redPoint, QPointF* greenPoint, QPointF* bluePoint, TransferFunction transferFunction) {
	return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, transferFunction);
}

QColorSpace* QColorSpace_new9(QPointF* whitePoint, QPointF* redPoint, QPointF* greenPoint, QPointF* bluePoint, struct miqt_array /* of uint16_t */  transferFunctionTable) {
	QList<uint16_t> transferFunctionTable_QList;
	transferFunctionTable_QList.reserve(transferFunctionTable.len);
	uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
	for(size_t i = 0; i < transferFunctionTable.len; ++i) {
		transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
	}
	return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new10(QPointF* whitePoint, QPointF* redPoint, QPointF* greenPoint, QPointF* bluePoint, struct miqt_array /* of uint16_t */  redTransferFunctionTable, struct miqt_array /* of uint16_t */  greenTransferFunctionTable, struct miqt_array /* of uint16_t */  blueTransferFunctionTable) {
	QList<uint16_t> redTransferFunctionTable_QList;
	redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
	uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
	for(size_t i = 0; i < redTransferFunctionTable.len; ++i) {
		redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> greenTransferFunctionTable_QList;
	greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
	uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
	for(size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
		greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> blueTransferFunctionTable_QList;
	blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
	uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
	for(size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
		blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
	}
	return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList);
}

QColorSpace* QColorSpace_new11(QColorSpace* colorSpace) {
	return new QColorSpace(*colorSpace);
}

QColorSpace* QColorSpace_new12(QPointF* whitePoint, TransferFunction transferFunction, float gamma) {
	return new QColorSpace(*whitePoint, transferFunction, static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new13(Primaries primaries, TransferFunction transferFunction, float gamma) {
	return new QColorSpace(primaries, transferFunction, static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new14(QPointF* whitePoint, QPointF* redPoint, QPointF* greenPoint, QPointF* bluePoint, TransferFunction transferFunction, float gamma) {
	return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, transferFunction, static_cast<float>(gamma));
}

void QColorSpace_OperatorAssign(QColorSpace* self, QColorSpace* colorSpace) {
	self->operator=(*colorSpace);
}

void QColorSpace_Swap(QColorSpace* self, QColorSpace* colorSpace) {
	self->swap(*colorSpace);
}

Primaries QColorSpace_Primaries(const QColorSpace* self) {
	return self->primaries();
}

TransferFunction QColorSpace_TransferFunction(const QColorSpace* self) {
	return self->transferFunction();
}

float QColorSpace_Gamma(const QColorSpace* self) {
	return self->gamma();
}

struct miqt_string QColorSpace_Description(const QColorSpace* self) {
	QString _ret = self->description();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QColorSpace_SetDescription(QColorSpace* self, struct miqt_string description) {
	QString description_QString = QString::fromUtf8(description.data, description.len);
	self->setDescription(description_QString);
}

void QColorSpace_SetTransferFunction(QColorSpace* self, TransferFunction transferFunction) {
	self->setTransferFunction(transferFunction);
}

void QColorSpace_SetTransferFunctionWithTransferFunctionTable(QColorSpace* self, struct miqt_array /* of uint16_t */  transferFunctionTable) {
	QList<uint16_t> transferFunctionTable_QList;
	transferFunctionTable_QList.reserve(transferFunctionTable.len);
	uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
	for(size_t i = 0; i < transferFunctionTable.len; ++i) {
		transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
	}
	self->setTransferFunction(transferFunctionTable_QList);
}

void QColorSpace_SetTransferFunctions(QColorSpace* self, struct miqt_array /* of uint16_t */  redTransferFunctionTable, struct miqt_array /* of uint16_t */  greenTransferFunctionTable, struct miqt_array /* of uint16_t */  blueTransferFunctionTable) {
	QList<uint16_t> redTransferFunctionTable_QList;
	redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
	uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
	for(size_t i = 0; i < redTransferFunctionTable.len; ++i) {
		redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> greenTransferFunctionTable_QList;
	greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
	uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
	for(size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
		greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> blueTransferFunctionTable_QList;
	blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
	uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
	for(size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
		blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
	}
	self->setTransferFunctions(redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList);
}

QColorSpace* QColorSpace_WithTransferFunction(const QColorSpace* self, TransferFunction transferFunction) {
	return new QColorSpace(self->withTransferFunction(transferFunction));
}

QColorSpace* QColorSpace_WithTransferFunctionWithTransferFunctionTable(const QColorSpace* self, struct miqt_array /* of uint16_t */  transferFunctionTable) {
	QList<uint16_t> transferFunctionTable_QList;
	transferFunctionTable_QList.reserve(transferFunctionTable.len);
	uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
	for(size_t i = 0; i < transferFunctionTable.len; ++i) {
		transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
	}
	return new QColorSpace(self->withTransferFunction(transferFunctionTable_QList));
}

QColorSpace* QColorSpace_WithTransferFunctions(const QColorSpace* self, struct miqt_array /* of uint16_t */  redTransferFunctionTable, struct miqt_array /* of uint16_t */  greenTransferFunctionTable, struct miqt_array /* of uint16_t */  blueTransferFunctionTable) {
	QList<uint16_t> redTransferFunctionTable_QList;
	redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
	uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
	for(size_t i = 0; i < redTransferFunctionTable.len; ++i) {
		redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> greenTransferFunctionTable_QList;
	greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
	uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
	for(size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
		greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
	}
	QList<uint16_t> blueTransferFunctionTable_QList;
	blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
	uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
	for(size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
		blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
	}
	return new QColorSpace(self->withTransferFunctions(redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList));
}

void QColorSpace_SetPrimaries(QColorSpace* self, Primaries primariesId) {
	self->setPrimaries(primariesId);
}

void QColorSpace_SetPrimaries2(QColorSpace* self, QPointF* whitePoint, QPointF* redPoint, QPointF* greenPoint, QPointF* bluePoint) {
	self->setPrimaries(*whitePoint, *redPoint, *greenPoint, *bluePoint);
}

void QColorSpace_SetWhitePoint(QColorSpace* self, QPointF* whitePoint) {
	self->setWhitePoint(*whitePoint);
}

QPointF* QColorSpace_WhitePoint(const QColorSpace* self) {
	return new QPointF(self->whitePoint());
}

TransformModel QColorSpace_TransformModel(const QColorSpace* self) {
	return self->transformModel();
}

ColorModel QColorSpace_ColorModel(const QColorSpace* self) {
	return self->colorModel();
}

void QColorSpace_Detach(QColorSpace* self) {
	self->detach();
}

bool QColorSpace_IsValid(const QColorSpace* self) {
	return self->isValid();
}

bool QColorSpace_IsValidTarget(const QColorSpace* self) {
	return self->isValidTarget();
}

QColorSpace* QColorSpace_FromIccProfile(struct miqt_string iccProfile) {
	QByteArray iccProfile_QByteArray(iccProfile.data, iccProfile.len);
	return new QColorSpace(QColorSpace::fromIccProfile(iccProfile_QByteArray));
}

struct miqt_string QColorSpace_IccProfile(const QColorSpace* self) {
	QByteArray _qb = self->iccProfile();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}

QColorTransform* QColorSpace_TransformationToColorSpace(const QColorSpace* self, QColorSpace* colorspace) {
	return new QColorTransform(self->transformationToColorSpace(*colorspace));
}

void QColorSpace_SetTransferFunction2(QColorSpace* self, TransferFunction transferFunction, float gamma) {
	self->setTransferFunction(transferFunction, static_cast<float>(gamma));
}

QColorSpace* QColorSpace_WithTransferFunction2(const QColorSpace* self, TransferFunction transferFunction, float gamma) {
	return new QColorSpace(self->withTransferFunction(transferFunction, static_cast<float>(gamma)));
}

void QColorSpace_Delete(QColorSpace* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QColorSpace*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qcolorspace.go
package qt

import (
	"unsafe"
)

type QColorSpace__NamedColorSpace int

const (
	QColorSpace__NamedColorSpace__SRgb        QColorSpace__NamedColorSpace = 1
	QColorSpace__NamedColorSpace__SRgbLinear  QColorSpace__NamedColorSpace = 2
	QColorSpace__NamedColorSpace__AdobeRgb    QColorSpace__NamedColorSpace = 3
	QColorSpace__NamedColorSpace__DisplayP3   QColorSpace__NamedColorSpace = 4
	QColorSpace__NamedColorSpace__ProPhotoRgb QColorSpace__NamedColorSpace = 5
	QColorSpace__NamedColorSpace__Bt2020      QColorSpace__NamedColorSpace = 6
	QColorSpace__NamedColorSpace__Bt2100Pq    QColorSpace__NamedColorSpace = 7
	QColorSpace__NamedColorSpace__Bt2100Hlg   QColorSpace__NamedColorSpace = 8
)

type QColorSpace__Primaries int

const (
	QColorSpace__Primaries__Custom      QColorSpace__Primaries = 0
	QColorSpace__Primaries__SRgb        QColorSpace__Primaries = 1
	QColorSpace__Primaries__AdobeRgb    QColorSpace__Primaries = 2
	QColorSpace__Primaries__DciP3D65    QColorSpace__Primaries = 3
	QColorSpace__Primaries__ProPhotoRgb QColorSpace__Primaries = 4
	QColorSpace__Primaries__Bt2020      QColorSpace__Primaries = 5
)

type QColorSpace__TransferFunction int

const (
	QColorSpace__TransferFunction__Custom      QColorSpace__TransferFunction = 0
	QColorSpace__TransferFunction__Linear      QColorSpace__TransferFunction = 1
	QColorSpace__TransferFunction__Gamma       QColorSpace__TransferFunction = 2
	QColorSpace__TransferFunction__SRgb        QColorSpace__TransferFunction = 3
	QColorSpace__TransferFunction__ProPhotoRgb QColorSpace__TransferFunction = 4
	QColorSpace__TransferFunction__Bt2020      QColorSpace__TransferFunction = 5
	QColorSpace__TransferFunction__St2084      QColorSpace__TransferFunction = 6
	QColorSpace__TransferFunction__Hlg         QColorSpace__TransferFunction = 7
)

type QColorSpace__TransformModel uint8_t

const (
	QColorSpace__ThreeComponentMatrix  QColorSpace__TransformModel = 0
	QColorSpace__ElementListProcessing QColorSpace__TransformModel = 1
)

type QColorSpace__ColorModel uint8_t

const (
	QColorSpace__Undefined QColorSpace__ColorModel = 0
	QColorSpace__Rgb       QColorSpace__ColorModel = 1
	QColorSpace__Gray      QColorSpace__ColorModel = 2
	QColorSpace__Cmyk      QColorSpace__ColorModel = 3
)

type QColorSpace struct {
	h          uintptr
	isSubclass bool
}

// NewQColorSpace constructs a new QColorSpace object.
func NewQColorSpace() *QColorSpace {
	g := newQColorSpace(QColorSpace_new())
	g.isSubclass = true
	return g
}

// NewQColorSpace2 constructs a new QColorSpace object.
func NewQColorSpace2(namedColorSpace NamedColorSpace) *QColorSpace {
	g := newQColorSpace(QColorSpace_new2(namedColorSpace))
	g.isSubclass = true
	return g
}

// NewQColorSpace3 constructs a new QColorSpace object.
func NewQColorSpace3(whitePoint QPointF, transferFunction TransferFunction) *QColorSpace {
	g := newQColorSpace(QColorSpace_new3(whitePoint.cPointer(), transferFunction))
	g.isSubclass = true
	return g
}

// NewQColorSpace4 constructs a new QColorSpace object.
func NewQColorSpace4(whitePoint QPointF, transferFunctionTable []uint16) *QColorSpace {
	transferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(transferFunctionTable))))
	defer free(unsafe.Pointer(transferFunctionTable_CArray))
	for i := range transferFunctionTable {
		transferFunctionTable_CArray[i] = (uint16_t)(transferFunctionTable[i])
	}
	transferFunctionTable_ma := struct_miqt_array{len: size_t(len(transferFunctionTable)), data: unsafe.Pointer(transferFunctionTable_CArray)}
	g := newQColorSpace(QColorSpace_new4(whitePoint.cPointer(), transferFunctionTable_ma))
	g.isSubclass = true
	return g
}

// NewQColorSpace5 constructs a new QColorSpace object.
func NewQColorSpace5(primaries Primaries, transferFunction TransferFunction) *QColorSpace {
	g := newQColorSpace(QColorSpace_new5(primaries, transferFunction))
	g.isSubclass = true
	return g
}

// NewQColorSpace6 constructs a new QColorSpace object.
func NewQColorSpace6(primaries Primaries, gamma float32) *QColorSpace {
	g := newQColorSpace(QColorSpace_new6(primaries, (float)(gamma)))
	g.isSubclass = true
	return g
}

// NewQColorSpace7 constructs a new QColorSpace object.
func NewQColorSpace7(primaries Primaries, transferFunctionTable []uint16) *QColorSpace {
	transferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(transferFunctionTable))))
	defer free(unsafe.Pointer(transferFunctionTable_CArray))
	for i := range transferFunctionTable {
		transferFunctionTable_CArray[i] = (uint16_t)(transferFunctionTable[i])
	}
	transferFunctionTable_ma := struct_miqt_array{len: size_t(len(transferFunctionTable)), data: unsafe.Pointer(transferFunctionTable_CArray)}
	g := newQColorSpace(QColorSpace_new7(primaries, transferFunctionTable_ma))
	g.isSubclass = true
	return g
}

// NewQColorSpace8 constructs a new QColorSpace object.
func NewQColorSpace8(whitePoint *QPointF, redPoint *QPointF, greenPoint *QPointF, bluePoint *QPointF, transferFunction TransferFunction) *QColorSpace {
	g := newQColorSpace(QColorSpace_new8(whitePoint.cPointer(), redPoint.cPointer(), greenPoint.cPointer(), bluePoint.cPointer(), transferFunction))
	g.isSubclass = true
	return g
}

// NewQColorSpace9 constructs a new QColorSpace object.
func NewQColorSpace9(whitePoint *QPointF, redPoint *QPointF, greenPoint *QPointF, bluePoint *QPointF, transferFunctionTable []uint16) *QColorSpace {
	transferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(transferFunctionTable))))
	defer free(unsafe.Pointer(transferFunctionTable_CArray))
	for i := range transferFunctionTable {
		transferFunctionTable_CArray[i] = (uint16_t)(transferFunctionTable[i])
	}
	transferFunctionTable_ma := struct_miqt_array{len: size_t(len(transferFunctionTable)), data: unsafe.Pointer(transferFunctionTable_CArray)}
	g := newQColorSpace(QColorSpace_new9(whitePoint.cPointer(), redPoint.cPointer(), greenPoint.cPointer(), bluePoint.cPointer(), transferFunctionTable_ma))
	g.isSubclass = true
	return g
}

// NewQColorSpace10 constructs a new QColorSpace object.
func NewQColorSpace10(whitePoint *QPointF, redPoint *QPointF, greenPoint *QPointF, bluePoint *QPointF, redTransferFunctionTable []uint16, greenTransferFunctionTable []uint16, blueTransferFunctionTable []uint16) *QColorSpace {
	redTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(redTransferFunctionTable))))
	defer free(unsafe.Pointer(redTransferFunctionTable_CArray))
	for i := range redTransferFunctionTable {
		redTransferFunctionTable_CArray[i] = (uint16_t)(redTransferFunctionTable[i])
	}
	redTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(redTransferFunctionTable)), data: unsafe.Pointer(redTransferFunctionTable_CArray)}
	greenTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(greenTransferFunctionTable))))
	defer free(unsafe.Pointer(greenTransferFunctionTable_CArray))
	for i := range greenTransferFunctionTable {
		greenTransferFunctionTable_CArray[i] = (uint16_t)(greenTransferFunctionTable[i])
	}
	greenTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(greenTransferFunctionTable)), data: unsafe.Pointer(greenTransferFunctionTable_CArray)}
	blueTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(blueTransferFunctionTable))))
	defer free(unsafe.Pointer(blueTransferFunctionTable_CArray))
	for i := range blueTransferFunctionTable {
		blueTransferFunctionTable_CArray[i] = (uint16_t)(blueTransferFunctionTable[i])
	}
	blueTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(blueTransferFunctionTable)), data: unsafe.Pointer(blueTransferFunctionTable_CArray)}
	g := newQColorSpace(QColorSpace_new10(whitePoint.cPointer(), redPoint.cPointer(), greenPoint.cPointer(), bluePoint.cPointer(), redTransferFunctionTable_ma, greenTransferFunctionTable_ma, blueTransferFunctionTable_ma))
	g.isSubclass = true
	return g
}

// NewQColorSpace11 constructs a new QColorSpace object.
func NewQColorSpace11(colorSpace *QColorSpace) *QColorSpace {
	g := newQColorSpace(QColorSpace_new11(colorSpace.cPointer()))
	g.isSubclass = true
	return g
}

// NewQColorSpace12 constructs a new QColorSpace object.
func NewQColorSpace12(whitePoint QPointF, transferFunction TransferFunction, gamma float32) *QColorSpace {
	g := newQColorSpace(QColorSpace_new12(whitePoint.cPointer(), transferFunction, (float)(gamma)))
	g.isSubclass = true
	return g
}

// NewQColorSpace13 constructs a new QColorSpace object.
func NewQColorSpace13(primaries Primaries, transferFunction TransferFunction, gamma float32) *QColorSpace {
	g := newQColorSpace(QColorSpace_new13(primaries, transferFunction, (float)(gamma)))
	g.isSubclass = true
	return g
}

// NewQColorSpace14 constructs a new QColorSpace object.
func NewQColorSpace14(whitePoint *QPointF, redPoint *QPointF, greenPoint *QPointF, bluePoint *QPointF, transferFunction TransferFunction, gamma float32) *QColorSpace {
	g := newQColorSpace(QColorSpace_new14(whitePoint.cPointer(), redPoint.cPointer(), greenPoint.cPointer(), bluePoint.cPointer(), transferFunction, (float)(gamma)))
	g.isSubclass = true
	return g
}

func (this *QColorSpace) OperatorAssign(colorSpace *QColorSpace) {
	QColorSpace_OperatorAssign(this.h, colorSpace.cPointer())
}

func (this *QColorSpace) Swap(colorSpace *QColorSpace) {
	QColorSpace_Swap(this.h, colorSpace.cPointer())
}

func (this *QColorSpace) Primaries() Primaries {
	xxxxxxxxx
}

func (this *QColorSpace) TransferFunction() TransferFunction {
	xxxxxxxxx
}

func (this *QColorSpace) Gamma() float32 {
	return (float32)(QColorSpace_Gamma(this.h))
}

func (this *QColorSpace) Description() string {
	var _ms struct_miqt_string = QColorSpace_Description(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QColorSpace) SetDescription(description string) {
	description_ms := struct_miqt_string{}
	description_ms.data = CString(description)
	description_ms.len = size_t(len(description))
	defer free(unsafe.Pointer(description_ms.data))
	QColorSpace_SetDescription(this.h, description_ms)
}

func (this *QColorSpace) SetTransferFunction(transferFunction TransferFunction) {
	QColorSpace_SetTransferFunction(this.h, transferFunction)
}

func (this *QColorSpace) SetTransferFunctionWithTransferFunctionTable(transferFunctionTable []uint16) {
	transferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(transferFunctionTable))))
	defer free(unsafe.Pointer(transferFunctionTable_CArray))
	for i := range transferFunctionTable {
		transferFunctionTable_CArray[i] = (uint16_t)(transferFunctionTable[i])
	}
	transferFunctionTable_ma := struct_miqt_array{len: size_t(len(transferFunctionTable)), data: unsafe.Pointer(transferFunctionTable_CArray)}
	QColorSpace_SetTransferFunctionWithTransferFunctionTable(this.h, transferFunctionTable_ma)
}

func (this *QColorSpace) SetTransferFunctions(redTransferFunctionTable []uint16, greenTransferFunctionTable []uint16, blueTransferFunctionTable []uint16) {
	redTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(redTransferFunctionTable))))
	defer free(unsafe.Pointer(redTransferFunctionTable_CArray))
	for i := range redTransferFunctionTable {
		redTransferFunctionTable_CArray[i] = (uint16_t)(redTransferFunctionTable[i])
	}
	redTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(redTransferFunctionTable)), data: unsafe.Pointer(redTransferFunctionTable_CArray)}
	greenTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(greenTransferFunctionTable))))
	defer free(unsafe.Pointer(greenTransferFunctionTable_CArray))
	for i := range greenTransferFunctionTable {
		greenTransferFunctionTable_CArray[i] = (uint16_t)(greenTransferFunctionTable[i])
	}
	greenTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(greenTransferFunctionTable)), data: unsafe.Pointer(greenTransferFunctionTable_CArray)}
	blueTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(blueTransferFunctionTable))))
	defer free(unsafe.Pointer(blueTransferFunctionTable_CArray))
	for i := range blueTransferFunctionTable {
		blueTransferFunctionTable_CArray[i] = (uint16_t)(blueTransferFunctionTable[i])
	}
	blueTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(blueTransferFunctionTable)), data: unsafe.Pointer(blueTransferFunctionTable_CArray)}
	QColorSpace_SetTransferFunctions(this.h, redTransferFunctionTable_ma, greenTransferFunctionTable_ma, blueTransferFunctionTable_ma)
}

func (this *QColorSpace) WithTransferFunction(transferFunction TransferFunction) *QColorSpace {
	_goptr := newQColorSpace(QColorSpace_WithTransferFunction(this.h, transferFunction))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) WithTransferFunctionWithTransferFunctionTable(transferFunctionTable []uint16) *QColorSpace {
	transferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(transferFunctionTable))))
	defer free(unsafe.Pointer(transferFunctionTable_CArray))
	for i := range transferFunctionTable {
		transferFunctionTable_CArray[i] = (uint16_t)(transferFunctionTable[i])
	}
	transferFunctionTable_ma := struct_miqt_array{len: size_t(len(transferFunctionTable)), data: unsafe.Pointer(transferFunctionTable_CArray)}
	_goptr := newQColorSpace(QColorSpace_WithTransferFunctionWithTransferFunctionTable(this.h, transferFunctionTable_ma))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) WithTransferFunctions(redTransferFunctionTable []uint16, greenTransferFunctionTable []uint16, blueTransferFunctionTable []uint16) *QColorSpace {
	redTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(redTransferFunctionTable))))
	defer free(unsafe.Pointer(redTransferFunctionTable_CArray))
	for i := range redTransferFunctionTable {
		redTransferFunctionTable_CArray[i] = (uint16_t)(redTransferFunctionTable[i])
	}
	redTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(redTransferFunctionTable)), data: unsafe.Pointer(redTransferFunctionTable_CArray)}
	greenTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(greenTransferFunctionTable))))
	defer free(unsafe.Pointer(greenTransferFunctionTable_CArray))
	for i := range greenTransferFunctionTable {
		greenTransferFunctionTable_CArray[i] = (uint16_t)(greenTransferFunctionTable[i])
	}
	greenTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(greenTransferFunctionTable)), data: unsafe.Pointer(greenTransferFunctionTable_CArray)}
	blueTransferFunctionTable_CArray := (*[0xffff]uint16_t)(malloc(size_t(8 * len(blueTransferFunctionTable))))
	defer free(unsafe.Pointer(blueTransferFunctionTable_CArray))
	for i := range blueTransferFunctionTable {
		blueTransferFunctionTable_CArray[i] = (uint16_t)(blueTransferFunctionTable[i])
	}
	blueTransferFunctionTable_ma := struct_miqt_array{len: size_t(len(blueTransferFunctionTable)), data: unsafe.Pointer(blueTransferFunctionTable_CArray)}
	_goptr := newQColorSpace(QColorSpace_WithTransferFunctions(this.h, redTransferFunctionTable_ma, greenTransferFunctionTable_ma, blueTransferFunctionTable_ma))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) SetPrimaries(primariesId Primaries) {
	QColorSpace_SetPrimaries(this.h, primariesId)
}

func (this *QColorSpace) SetPrimaries2(whitePoint *QPointF, redPoint *QPointF, greenPoint *QPointF, bluePoint *QPointF) {
	QColorSpace_SetPrimaries2(this.h, whitePoint.cPointer(), redPoint.cPointer(), greenPoint.cPointer(), bluePoint.cPointer())
}

func (this *QColorSpace) SetWhitePoint(whitePoint QPointF) {
	QColorSpace_SetWhitePoint(this.h, whitePoint.cPointer())
}

func (this *QColorSpace) WhitePoint() *QPointF {
	_goptr := newQPointF(QColorSpace_WhitePoint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) TransformModel() TransformModel {
	xxxxxxxxx
}

func (this *QColorSpace) ColorModel() ColorModel {
	xxxxxxxxx
}

func (this *QColorSpace) Detach() {
	QColorSpace_Detach(this.h)
}

func (this *QColorSpace) IsValid() bool {
	return (bool)(QColorSpace_IsValid(this.h))
}

func (this *QColorSpace) IsValidTarget() bool {
	return (bool)(QColorSpace_IsValidTarget(this.h))
}

func QColorSpace_FromIccProfile(iccProfile []byte) *QColorSpace {
	iccProfile_alias := struct_miqt_string{}
	iccProfile_alias.data = (char)(unsafe.Pointer(&iccProfile[0]))
	iccProfile_alias.len = size_t(len(iccProfile))
	_goptr := newQColorSpace(QColorSpace_FromIccProfile(iccProfile_alias))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) IccProfile() []byte {
	var _bytearray struct_miqt_string = QColorSpace_IccProfile(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QColorSpace) TransformationToColorSpace(colorspace *QColorSpace) *QColorTransform {
	_goptr := newQColorTransform(QColorSpace_TransformationToColorSpace(this.h, colorspace.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QColorSpace) SetTransferFunction2(transferFunction TransferFunction, gamma float32) {
	QColorSpace_SetTransferFunction2(this.h, transferFunction, (float)(gamma))
}

func (this *QColorSpace) WithTransferFunction2(transferFunction TransferFunction, gamma float32) *QColorSpace {
	_goptr := newQColorSpace(QColorSpace_WithTransferFunction2(this.h, transferFunction, (float)(gamma)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}
